h2o-algos/src/main/java/hex/word2vec/WordVectorTrainer.java

```diff
@@ -131,200 +87,111 @@ private void initExpTable() {
-      } // while more sentences
-    }
-    _output.addLocallyProcessed(wrdCnt);
-  }
-
-  @Override public void reduce (WordVectorTrainer other) {
-    if (other._output.getLocallyProcessed() > 0 //other task was active (its syn0 should be used for averaging)
-            && other._output != _output) //other task worked on a different syn0
-    {
-      // avoid adding remote model info to unprocessed local data
-      // (can happen if master node has no chunks)
-      if (_output.getLocallyProcessed() == 0) {
-        _output = other._output;
-        _chunkNodeCount = other._chunkNodeCount;
-      } else {
-        _output.add(other._output);
-        _chunkNodeCount += other._chunkNodeCount;
-      }
+        } // end for each item in the window
+
+        wordCount++;
+        // update learning rate
+        if (wordCount % 10000 == 0) {
+          _nodeProcessedWords._val += 10000;
+          long totalProcessedWordsEst = _prevTotalProcessedWords + _nodeProcessedWords._val;
+          _curLearningRate = calcLearningRate(_initLearningRate, _epochs, totalProcessedWordsEst, _vocabWordCount);
+        }
+      } // for each item in the sentence
+    } // while more sentences
+    _processedWords = wordCount;
+    _nodeProcessedWords._val += wordCount % 10000;
+  }
+
+  @Override public void reduce(WordVectorTrainer other) {
+    _processedWords += other._processedWords;
+    if (_syn0 != other._syn0) { // other task worked on a different syn0
+      float c = (float) other._processedWords / _processedWords;
+      ArrayUtils.add(1.0f - c, _syn0, c, other._syn0);
+      ArrayUtils.add(1.0f - c, _syn1, c, other._syn1);
+      // for diagnostics only
+      _nodeProcessedWords._val += other._nodeProcessedWords._val;
     }
   }
 
-  @Override
-  protected void closeLocal() {
-    _vocab = null;
-  }
-
-  static long _lastWarn, _warnCount;
-  @Override protected void postGlobal(){
-    if (H2O.CLOUD.size() > 1) {
-      long now = System.currentTimeMillis();
-      if (_chunkNodeCount < H2O.CLOUD.size() && (now - _lastWarn > 5000) && _warnCount < 3) {
-        Log.warn(H2O.CLOUD.size() - _chunkNodeCount + " node(s) (out of " + H2O.CLOUD.size()
-                + ") are not contributing to model updates. Consider setting replicate_training_data to true or using a larger training dataset (or fewer H2O nodes).");
-        _lastWarn = now;
-        _warnCount++;
-      }
-    }
-    _output.div(_chunkNodeCount);
-    _output.addGloballyProcessed(_output.getLocallyProcessed());
-    _output.setLocallyProcessed(0);
-
-    assert(_input == null);
-  }
-
   private void skipGram(int curWord, int winWord, float[] neu1e) {
     final int vecSize = _wordVecSize;
     final int l1 = winWord * vecSize;
     for (int i = 0; i < vecSize; i++) neu1e[i] = 0;
 
-    if (_normModel == NormModel.NegSampling)
-      negSamplingSG(curWord, l1, neu1e);
-    else // HSM
-      hierarchicalSoftmaxSG(curWord, l1, neu1e);
+    hierarchicalSoftmaxSG(curWord, l1, neu1e);
```

Comment:
Yes, I wanted to keep it simple for the first version. Adding NS (and CBOW) should be relatively easy in next iterations. For now I think it is okay to just do Skip-Gram with Hierarchical Softmax (that is what Spark has).